service/worker/scanner/tasklist/handler.go

```diff
@@ -47,60 +48,45 @@ const scannerTaskListPrefix = "cadence-sys-tl-scanner"
 // with the assumption that the executor will schedule this task later
 //
 // Each loop of the handler proceeds as follows
-//    - Retrieve the next batch of tasks sorted by task_id for this task-list from persistence
+//    - Attempt to delete tasks up to the persisted ACK level.
```

Comment:
I can't say with certainty that infinite would be stored, but I saw in direct observation that we had tasks stored which had expiration dates more than 300 years in the future. That effectively neutered scavenger.

Reply:
Sorry, just want to make myself clear here, in case there's any misunderstanding. 
- I am not against the idea of deleting acked tasks in scanner, as it can cover some rare cases where acked tasks are not deleted in time. 
- **Regardless of a task is acked or not, as long as it has passed the expiration, it needs to be deleted, even if it has a taskID higher than the ack level.** This is the main purpose of task list scanner since there're cases where a task may never be acked.

Reply:
I am not sure where the 300 year expiration comes from. From design perspective it should not happen and there should be a 1 year limit for the expiration in history service when pushing tasks to matching. If you do see an unreasonable expiration interval, it must be a bug somewhere in the system and we need to fix it.

Reply:
Thanks for the confirmation about deleting expired tasks.

I'm not sure on the long expiration either. I am going to add some logging to see if I may get to the bottom of it. (It's difficult to directly search for such tasks since the expiration date is "hidden" in the binary data from a db query point of view.) One thought I had on the topic; if a retry policy is set its duration will override the start-to-close timeout. Maybe it's possible to set a retry policy that implies a very long timeout and thus bypass the limits?

Reply:
I still see unreasonable expiration times on a handful of tasks. Now that I'm logging them, I can see they have the following properties:
* The date stored in the DB is the value 11651379494838206464 which convert to Tuesday, March 21, 2339 10:18:14.838 PM
* This actually is interpreted as 1754-08-30 22:43:41.128654848 +0000 UTC presumably due to overflow when converting from 64 bit epoch nanoseconds to 32 bit epoch seconds.
* However isTaskExpired also checks for t.Expiry.After(epochStartTime) so it still treats this negative time as unexpired. The presence of the check makes me think these overflow values are not a new surprise.

All of these tasks have identically the same expiration value stored in the database, which makes me think it must have been statically initialized from some value.

Reply:
Can you file another bug/issue with more information so that we can fix it separately?:
* server version
* tasklist type(decision/activity)
* tasklist kind(sticky/non-stikcy)
* the deserialized taskInfo object so that we can see `workflowID,runID, scheduledID, expiryTime and createdTime`